pkg/ck8s/workload_cluster.go

```diff
@@ -192,58 +206,93 @@ func (w *Workload) GetK8sdProxyForControlPlane(ctx context.Context) (*K8sdClient
 // NewWorkerJoinToken reaches out to the control-plane of the workload cluster via k8sd-proxy client.
-func (w *Workload) NewWorkerJoinToken(ctx context.Context, authToken string, microclusterPort int, name string) (string, error) {
-	return w.requestJoinToken(ctx, microclusterPort, authToken, name, true)
+func (w *Workload) NewWorkerJoinToken(ctx context.Context, authToken string, name string) (string, error) {
+	return w.requestJoinToken(ctx, authToken, name, true)
 }
 
 // requestJoinToken requests a join token from the existing control-plane nodes via the k8sd proxy.
-func (w *Workload) requestJoinToken(ctx context.Context, microclusterPort int, authToken string, name string, worker bool) (string, error) {
-	// FIXME: We need to check the default value for this port in one place, not in multiple places.
-	// See https://github.com/canonical/cluster-api-k8s/pull/8#discussion_r1645536362 for context.
-	if microclusterPort == 0 {
-		microclusterPort = 2380
+func (w *Workload) requestJoinToken(ctx context.Context, authToken string, name string, worker bool) (string, error) {
+	request := apiv1.GetJoinTokenRequest{Name: name, Worker: worker}
+	response := &apiv1.GetJoinTokenResponse{}
+	err := w.doK8sdRequest(ctx, http.MethodPost, "1.0/x/capi/generate-join-token", authToken, request, response, k8sdProxyOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get join token: %w", err)
 	}
+	return response.EncodedToken, nil
+}
 
-	k8sdProxy, err := w.GetK8sdProxyForControlPlane(ctx)
+func (w *Workload) RemoveMachineFromCluster(ctx context.Context, machine *clusterv1.Machine, authToken string) error {
+	request := &apiv1.RemoveNodeRequest{Name: machine.Name, Force: true}
+
+	// If we see that blacklisting control-planes is causing issues, let's consider removing it.
+	// It *should* not be necessary as a machine should be able to remove itself from the cluster.
+	err := w.doK8sdRequest(ctx, http.MethodPost, "1.0/x/capi/remove-node", authToken, request, nil, k8sdProxyOptions{BlacklistedControlPlanes: []string{machine.Name}})
 	if err != nil {
-		return "", fmt.Errorf("failed to create k8sd proxy: %w", err)
+		return fmt.Errorf("failed to remove %s from cluster: %w", machine.Name, err)
 	}
+	return nil
+}
+
+func (w *Workload) doK8sdRequest(ctx context.Context, method, endpoint, authToken string, request any, response any, k8sdProxyOptions k8sdProxyOptions) error {
 	type wrappedResponse struct {
-		Error    string                     `json:"error"`
-		Metadata apiv1.GetJoinTokenResponse `json:"metadata"`
+		Error    string          `json:"error"`
+		Metadata json.RawMessage `json:"metadata"`
 	}
 
-	requestBody, err := json.Marshal(apiv1.GetJoinTokenRequest{Name: name, Worker: worker})
+	// FIXME: Get this from the workload cluster's configuration instead of hardcoding it.
+	// See https://github.com/canonical/cluster-api-k8s/pull/8#discussion_r1645536362 for context.
+	microclusterPort := 2380
```

Comment:
please, let's take this in when calling `GetWorkloadCluster()`, do not hardcode this.